src/include/libpmemobj++/persistent_ptr.hpp

```diff
@@ -217,78 +143,23 @@ class persistent_ptr {
 	typename nvml::detail::sp_member_access<T>::type operator->() const
 		noexcept
 	{
-		return get();
+		return this->get();
 	}
 
 	/**
 	 * Array access operator.
 	 *
 	 * Contains run-time bounds checking for static arrays.
 	 */
+	template <typename = typename std::enable_if<!std::is_void<T>::value>>
 	typename nvml::detail::sp_array_access<T>::type
 	operator[](std::ptrdiff_t i) const noexcept
 	{
 		assert(i >= 0 && (i < nvml::detail::sp_extent<T>::value ||
```

Comment:
The thing is, the allocator does not make sense if it does not work nicely with standard library containers and containers need pointers. One special kind is a `void` pointer which has rules of its own. I probably could separate this into two commits, is that what you reeeeaaaly want?